### ISSUE TITLE
RepositoryWebHook creation fails

### ISSUE THREAD

**Issue by nosedrum** [ovh/cds#5356]
Hello,

When I add a RepositoryWebHook (bitbucket cloud synced) to a pipeline, the creation crash with error on api log:

> `2020-07-31 14:31:48 [INFO] pprofLabel>compress>setRequestID>recoverWrap>handle>putWorkflowHandler>Update>hookRegistration>createVCSConfiguration: Invalid hook configuration (caused by: cannot update workflow: given webhook url value /webhook/b0cb67ab-83f9-4e92-9ccf-61f83a2c508e is not a url) auth_consumer_id=33636938-751c-460a-bcad-b950f75912ae auth_session_id=3fb6269a-ce10-4e54-9cae-fec90c13cfe4 auth_user_id=c1beb9e4-fc70-4774-895f-b443d66ecf3e request_id=ef7fb441-50d3-4008-b157-7fc3a0c27dd7 stack_trace=given webhook url value /webhook/b0cb67ab-83f9-4e92-9ccf-61f83a2c508e is not a url
github.com/ovh/cds/sdk.WrapError
        /tmp/QnVpbGQgYW5kIFBhY2thZ2UgQWxs/run/sdk/error.go:807
github.com/ovh/cds/engine/api/workflow.createVCSConfiguration
        /tmp/QnVpbGQgYW5kIFBhY2thZ2UgQWxs/run/engine/api/workflow/hook.go:309
github.com/ovh/cds/engine/api/workflow.hookRegistration
        /tmp/QnVpbGQgYW5kIFBhY2thZ2UgQWxs/run/engine/api/workflow/hook.go:195
github.com/ovh/cds/engine/api/workflow.Update
        /tmp/QnVpbGQgYW5kIFBhY2thZ2UgQWxs/run/engine/api/workflow/dao.go:621
github.com/ovh/cds/engine/api.(*API).putWorkflowHandler.func1
        /tmp/QnVpbGQgYW5kIFBhY2thZ2UgQWxs/run/engine/api/workflow.go:444
github.com/ovh/cds/engine/api.(*Router).handle.func1
        /tmp/QnVpbGQgYW5kIFBhY2thZ2UgQWxs/run/engine/api/router.go:442
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2012
github.com/ovh/cds/engine/api.(*Router).recoverWrap.func1
        /tmp/QnVpbGQgYW5kIFBhY2thZ2UgQWxs/run/engine/api/router.go:177
github.com/ovh/cds/engine/api.(*Router).setRequestID.func1
        /tmp/QnVpbGQgYW5kIFBhY2thZ2UgQWxs/run/engine/api/router.go:127
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2012
github.com/gorilla/handlers.CompressHandlerLevel.func1
        /go/pkg/mod/github.com/gorilla/handlers@v0.0.0-20160816184729-a5775781a543/compress.go:143
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2012
github.com/ovh/cds/engine/api.(*Router).compress.func1
        /tmp/QnVpbGQgYW5kIFBhY2thZ2UgQWxs/run/engine/api/router.go:100
github.com/ovh/cds/engine/api.(*Router).pprofLabel.func1
        /tmp/QnVpbGQgYW5kIFBhY2thZ2UgQWxs/run/engine/api/router.go:92
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2012
github.com/gorilla/mux.(*Router).ServeHTTP
        /go/pkg/mod/github.com/gorilla/mux@v1.6.2/mux.go:162
net/http.serverHandler.ServeHTTP
        /usr/local/go/src/net/http/server.go:2807
net/http.(*conn).serve
        /usr/local/go/src/net/http/server.go:1895
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1373
cannot update workflow`

All seems ok on vcs logs.
Project is linked with a bitbucket cloud repository manager (works perfectly with checkout in pipeline and payload for workflow launch)
Application is linked to the right git repository (works well in pipeline)
Repository Webhook is attempted to be added on the root pipeline and this pipeline have the application linked in the context.

What is wrong ? Any Idea ?

**Comment by nosedrum** [comment]
Never mind "url" in [hooks] wasn't specified in conf.toml...

**Comment by yesnault** [comment]
`./engine config new hooks` will display the configuration for hooks service. The urlPublic is sent to bitbucket, to let it to call your hook service on each git push for example. CDS need it to configure a new repositoryWebhook, urlPublic should be mandatory. I reopen this ticket for now.

**Comment by nosedrum** [comment]
With "urlPublic" repositoryWebhook on repo:push event  from bitbucket cloud is created ok.
But, when I push on branch "toto" the pipeline run with git.branch = master !
I try to play around the default payload with UI and as code. Nothing to do, the Hook trigger always run pipeline with git.branch=master ...
No erros in any logs.
I check the body of bitbucket cloud request and all is ok inside, the good branch "toto" is specified.
I was thinking, git.branch was automaticaly setup with the branch name where we push. 
Do I miss something in repositoryWebhook parameters for automatic git.branch setting ?

**Comment by yesnault** [comment]
yes, it's now another subject.

git.branch is getted from payload if you run manually.
if bitbucket triggers the run, the payload is completed with data from the call from bitbucket.
otherwise, git.branch variable is computed by the CheckoutApplication / GitClone actions.

Where do you see that the pipeline is running with git.branch = master? On 'Display Job Variables' or do you try to enter the a run condition ?

**Comment by nosedrum** [comment]
> git.branch is getted from payload if you run manually.

- yes, manually it's working perfectly.

> if bitbucket triggers the run, the payload is completed with data from the call from bitbucket.

- It's my case. On a push, bitbucket cloud send a request to my vcs. In the bitbucket cloud payload, the good branch ("toto" ie). The request is receive, the run triggered, but in first step I do `echo {{.git.branch}}` to check and it's `master`. When I check in 'Display Job Variables' on pipeline run logs history, it's confirmed git.branch = master.

Here a sample of bitbucket cloud request body: 
`{
  "push": {
    "changes": [
      {
        "forced": false,
        "old": {
          "name": "toto",
          "links": {
            "commits": {
              "href": "https://api.bitbucket.org/2.0/repositories/xxxxxxxx/xxxxxx/commits/toto"
            },
            "self": {
              "href": "https://api.bitbucket.org/2.0/repositories/xxxxxxxx/xxxxxx/refs/branches/toto"
            },
            "html": {
              "href": "https://bitbucket.org/xxxxxxxx/xxxxxx/branch/toto"
            }`
...
` "new": {
          "name": "toto",
          "links": {
            "commits": {
              "href": "https://api.bitbucket.org/2.0/repositories/xxxxxxxx/xxxxxx/commits/toto"
            },
            "self": {
              "href": "https://api.bitbucket.org/2.0/repositories/xxxxxxxx/xxxxxx/refs/branches/toto"
            },
            "html": {
              "href": "https://bitbucket.org/xxxxxxxx/xxxxxx/branch/toto"
            }`

> otherwise, git.branch variable is computed by the CheckoutApplication / GitClone actions.

- I use a CheckoutApplication action on start of my pipeline triggered by bbc, but I can specify only a directory, I don't see any 'branch' variable. 
- In GitClone action there is a branch variable but how to force the name of the branch where the push occured? with {{.git.branch}} ? (I'll try this)

> Where do you see that the pipeline is running with git.branch = master? On 'Display Job Variables' or do you try to enter the a run condition ?
- On 'Display Job Variables', yes, has I said, in pipeline exection logs is equal to master systematicaly.
- If I try to enter as a run condition, the pipeline don't run...

**Comment by nosedrum** [comment]
To complete, I just test with GitClone action:

```
branch = {{.git.branch}}
commit = {{.git.hash}}
directory = {{.cds.workspace}}
password = xxxxxx
privateKey = null
url = {{.git.http_url}}
user = xxxxxxxxxx
```

Result, again 'master' ! :

```
[INFO] Starting step "GitClone"
[INFO] Executing: git clone --depth 50 --branch master --recursive...
Executing: git reset --hard 1fd9df91c204ef9aa767a693a9f83f0becaa64bf
[INFO] HEAD is now at 1fd9df9 xxxxxxxxxxxxxxxxxxxxxxx
[WARN] Cloning into '/tmp/QnVpbGQgYW5kIFRlc3Q/run'...
[INFO] git.branch: master
[INFO] git.hash: 1fd9df91c204ef9aa767a693a9f83f0becaa64bf
[INFO] git.message: xxxxxxxxxxxxxxxxxxxxxxxxx
[INFO] git.author: Xxxxxxxxx Xxxxxxxxx
[INFO] git.author.email: xxxxx@xxxxx.xxx
[INFO] End of step "GitClone" (6s)
```

**Comment by nosedrum** [comment]
any advice ?

**Comment by yesnault** [comment]
Hi @nosedrum , we should check that this week

**Comment by yesnault** [comment]
@nosedrum fix incoming, it will available in next release 0.48.0
